File: trunk/Collective/src/collective/agents/QuadTree.scala
package collective.agents

//import agents.Obstacle

class QuadTree {
  var obstacles = Nil //new List[Obstacle]
  
  def contains(x: Int,y: Int):Boolean =
  {
		  true
  }
  def add(x: Int,y: Int):Boolean  = //returns true if successful
  {
	  true
  }
  def remove(x: Int,y: Int):Boolean = //returns true if successful
  {
	  true
  }
  def range(sensorRange: Int, x: Int, y: Int):List[Obstacle] = 
  {
		  for (
				  obstacle: Obstacle <- obstacles
				  if(obstacle.x*obstacle.x* + obstacle.y*obstacle.y <= sensorRange)
		  ) yield {
			  obstacle
		  }

  }

	


}

File: trunk/Collective/src/collective/agents/QuadTree.java
package collective.agents;

public class QuadTree{
  QuadNode head = null;
  int length = 0;
  public QuadTree(){
    
  }
  
  public QuadTree(Object o){
    head = new QuadNode(o);
    length = 1;
  }
  
  public QuadTree(Object[] os){
    for(int i = os.length-1; i >= 0; i++){
      QuadNode added = new QuadNode(os[i]);
      added.next = head;
    }
    length = os.length;
  }
  
  public void add(Object o){
	  addAtBeg(o);
  }
  
  public void addAtBeg(Object o){
    QuadNode added = new QuadNode(o);
    added.next = head;
    length++;
  }
  
  public Object find(Object o){
    QuadNode ln = head;
    while(ln.next!=null){
    	ln = ln.next;
    	if (ln.obj.toString().equals(o.toString()))
        	return o;
    }
    if(ln.obj.toString().equals(o.toString()))
    	return o;
    return null;
  }
  
  public Object[] returnAll(){
	  Object[] op = new Object[length];
	  QuadNode ln = head;
	  for(int i = 0; i < length; i++){
		  op[i]=ln;
		  ln = ln.next;
	  }
	  return op;
  }
  
  public int remove(Object o){
	  QuadNode ln = head;
	  if(ln.obj.toString().equals(o.toString())){
		  head = ln.next;
		  ln.next = null;
		  return 1;
	  }
	  while(ln.next!=null){
	    if(ln.next.obj.toString().equals(o.toString())){
	    	ln.next = ln.next.next;
	    	return 1;
	    }
	    ln = ln.next;	
	  }
	  return 0;
  }
}
